modules/proposals/engine/src/lib.rs

```diff
@@ -0,0 +1,348 @@
+        ensure!(
+            body.len() as u32 <= Self::body_max_len(),
+            errors::MSG_TOO_LONG_BODY
+        );
+
+        let next_proposal_count_value = Self::proposal_count() + 1;
+        let new_proposal_id = next_proposal_count_value;
+
+        let new_proposal = Proposal {
+            created: Self::current_block(),
+            parameters,
+            title,
+            body,
+            proposer_id: proposer_id.clone(),
+            proposal_type,
+            status: ProposalStatus::Active,
+        };
+
+        // mutation
+        <Proposals<T>>::insert(new_proposal_id, new_proposal);
+        <ProposalCode>::insert(new_proposal_id, proposal_code);
+        ActiveProposalIds::mutate(|ids| ids.insert(new_proposal_id));
+        ProposalCount::put(next_proposal_count_value);
+
+        Self::deposit_event(RawEvent::ProposalCreated(proposer_id, new_proposal_id));
+
+        Ok(())
+    }
+}
+
+impl<T: Trait> Module<T> {
+    // Wrapper-function over system::block_number()
+    fn current_block() -> T::BlockNumber {
+        <system::Module<T>>::block_number()
+    }
+
+    // Executes approved proposal code
+    fn execute_proposal(proposal_id: u32) {
+        //let origin = system::RawOrigin::Root.into();
+        let proposal = Self::proposals(proposal_id);
+        let proposal_code = Self::proposal_codes(proposal_id);
+
+        let proposal_code_result =
+            T::ProposalCodeDecoder::decode_proposal(proposal.proposal_type, proposal_code);
+
+        let new_proposal_status = match proposal_code_result {
+            Ok(proposal_code) => {
+                if let Err(error) = proposal_code.execute() {
+                    ProposalStatus::Failed {
+                        error: error.as_bytes().to_vec(),
+                    }
+                } else {
+                    ProposalStatus::Executed
+                }
+            }
+            Err(error) => ProposalStatus::Failed {
+                error: error.as_bytes().to_vec(),
+            },
+        };
+
+        Self::update_proposal_status(proposal_id, new_proposal_status)
+    }
+
+    /// Voting results tally.
+    /// Returns proposals with changed status and tally results
+    fn tally() -> Vec<TallyResult<T::BlockNumber>> {
+        let mut results = Vec::new();
+        for &proposal_id in Self::active_proposal_ids().iter() {
+            let votes = Self::votes_by_proposal(proposal_id);
+            let proposal = Self::proposals(proposal_id);
+
+            if let Some(tally_result) = proposal.tally_results(
+                proposal_id,
+                votes,
+                T::TotalVotersCounter::total_voters_count(),
+                Self::current_block(),
+            ) {
+                results.push(tally_result);
+            }
+        }
+
+        results
+    }
+
+    /// Updates proposal status and removes proposal id from active id set.
+    fn update_proposal_status(proposal_id: u32, new_status: ProposalStatus) {
```

Comment:
Its very hard to reason about the correctness of this routine, at least on first reading. It is called in so many different scenarios, and its unclear what invariant must hold for each cod path that has side effects here. Can we avoid this construction? The fact that there is an unconditional removal from `ActiveProposalIds` seems at least semantically incorrect, e.g when `new_status == ProposalStatus::Active`.

Reply:
The update_proposal_status() called in many scenarios because of my intention to have a single status transition place. Also, it has no invariant checks because we have not addressed this issue before Feb 12. Now I can forbid invalid transitions. By the way, should I?

Reply:
Decided to postpone what to do with this and the corresponding general `ProposalStatusUpdated` event at a later stage.

Reply:
Marked with TODO. I'm going to refactor or get rid of it after adding stakes.